bn256.G2 points fail on in-place addition

Note that this is not a security issue, since `bn256.G1.Add` is only used in `vm.runBn256Add`, and is not used in-place, there. Also, a [fix has been provided upstream](https://github.com/cloudflare/bn256/pull/12), and just hasn't been pulled in. My recommendation to fix this would be to make [these changes](https://github.com/cloudflare/bn256/pull/12/files) to to `go-ethereum`'s `cloudflare/bn256`.

A failing test is provided below, in the "Expected Behavior" section.

#### System information

Geth `version:` `geth version` => `1.15` (master) 
OS & Version: Windows/Linux/OSX
```
uname -a
Linux ubuntu-VirtualBox 5.4.0-74-generic #83-Ubuntu SMP Sat May 8 02:35:39 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```

Commit hash : (if `develop`) `28ba686cb`

#### Expected behaviour

Following test of `bn256.G1` addition should pass:

```golang
func TestG2SelfAddition(t *testing.T) {
	s, _ := rand.Int(rand.Reader, Order)
	p := new(G2).ScalarBaseMult(s)
	if _, err := p.Unmarshal(p.Add(p, p).Marshal()); err != nil {
		t.Fatalf("p.Add(p, p) ∉ G₂: %v", err)
	}
}
```

#### Actual behaviour

```
ubuntu@ubuntu-VirtualBox:~/alien/go-ethereum/crypto/bn256/cloudflare$ go test -run  TestG2SelfAddition
--- FAIL: TestG2SelfAddition (0.00s)
    bn256_test.go:122: p.Add(p, p) ∉ G₂: bn256: coordinate exceeds modulus
FAIL
exit status 1
FAIL	github.com/ethereum/go-ethereum/crypto/bn256/cloudflare	0.005s
```

#### Steps to reproduce the behaviour

- Add the above test to `crypto/bn256/cloudflare/bn256_test.go` on `master`
- `go test -run  TestG2SelfAddition`

#### Backtrace

N/A

--- a/crypto/bn256/cloudflare/bn256_test.go
+++ b/crypto/bn256/cloudflare/bn256_test.go
@@ -92,6 +92,19 @@ func TestTripartiteDiffieHellman(t *testing.T) {
 	}
 }
 
+func TestG2SelfAddition(t *testing.T) {
+	s, _ := rand.Int(rand.Reader, Order)
+	p := new(G2).ScalarBaseMult(s)
+
+	if !p.p.IsOnCurve() {
+		t.Fatal("p isn't on curve")
+	}
+	m := p.Add(p, p).Marshal()
+	if _, err := p.Unmarshal(m); err != nil {
+		t.Fatalf("p.Add(p, p) ∉ G₂: %v", err)
+	}
+}
+
 func BenchmarkG1(b *testing.B) {
 	x, _ := rand.Int(rand.Reader, Order)
 	b.ResetTimer()

--- a/crypto/bn256/cloudflare/curve.go
+++ b/crypto/bn256/cloudflare/curve.go
@@ -171,15 +171,15 @@ func (c *curvePoint) Double(a *curvePoint) {
 	gfpAdd(t, d, d)
 	gfpSub(&c.x, f, t)
 
+	gfpMul(&c.z, &a.y, &a.z)
+	gfpAdd(&c.z, &c.z, &c.z)
+
 	gfpAdd(t, C, C)
 	gfpAdd(t2, t, t)
 	gfpAdd(t, t2, t2)
 	gfpSub(&c.y, d, &c.x)
 	gfpMul(t2, e, &c.y)
 	gfpSub(&c.y, t2, t)
-
-	gfpMul(t, &a.y, &a.z)
-	gfpAdd(&c.z, t, t)
 }
 
 func (c *curvePoint) Mul(a *curvePoint, scalar *big.Int) {

--- a/crypto/bn256/cloudflare/gfp.go
+++ b/crypto/bn256/cloudflare/gfp.go
@@ -61,6 +61,7 @@ func (e *gfP) Marshal(out []byte) {
 func (e *gfP) Unmarshal(in []byte) error {
 	// Unmarshal the bytes into little endian form
 	for w := uint(0); w < 4; w++ {
+		e[3-w] = 0
 		for b := uint(0); b < 8; b++ {
 			e[3-w] += uint64(in[8*w+b]) << (56 - 8*b)
 		}

--- a/crypto/bn256/cloudflare/twist.go
+++ b/crypto/bn256/cloudflare/twist.go
@@ -150,15 +150,15 @@ func (c *twistPoint) Double(a *twistPoint) {
 	t.Add(d, d)
 	c.x.Sub(f, t)
 
+	c.z.Mul(&a.y, &a.z)
+	c.z.Add(&c.z, &c.z)
+
 	t.Add(C, C)
 	t2.Add(t, t)
 	t.Add(t2, t2)
 	c.y.Sub(d, &c.x)
 	t2.Mul(e, &c.y)
 	c.y.Sub(t2, t)
-
-	t.Mul(&a.y, &a.z)
-	c.z.Add(t, t)
 }
 
 func (c *twistPoint) Mul(a *twistPoint, scalar *big.Int) {

